src/Microsoft.Data/StringColumn.cs

```diff
@@ -273,26 +271,76 @@ private StringColumn Clone(PrimitiveColumn<bool> boolColumn)
             return ret;
         }
 
-        private StringColumn CloneImplementation(BaseColumn mapIndices, Func<long, long?> getIndex, bool invertMapIndices = false)
+        private StringColumn CloneImplementation<U>(PrimitiveColumn<U> mapIndices, bool invertMapIndices = false)
+            where U : unmanaged
         {
-            StringColumn ret = new StringColumn(Name, mapIndices is null ? Length : mapIndices.Length);
-            if (mapIndices.Length > Length)
-                throw new ArgumentException(Strings.MapIndicesExceedsColumnLenth, nameof(mapIndices));
-            if (invertMapIndices == false)
+            mapIndices = mapIndices ?? throw new ArgumentNullException(nameof(mapIndices));
+            StringColumn ret = new StringColumn(Name, mapIndices.Length);
+
+            List<string> setBuffer = ret._stringBuffers[0];
+            long setBufferMinRange = 0;
+            long setBufferMaxRange = int.MaxValue;
+            List<string> getBuffer = _stringBuffers[0];
+            long getBufferMinRange = 0;
+            long getBufferMaxRange = int.MaxValue;
+            long maxCapacity = int.MaxValue;
+            if (mapIndices.DataType == typeof(long))
             {
-                for (long i = 0; i < mapIndices.Length; i++)
+                PrimitiveColumn<long> longMapIndices = mapIndices as PrimitiveColumn<long>;
+                longMapIndices.ApplyElementwise((long? mapIndex, long rowIndex) =>
                 {
-                    ret[i] = this[getIndex(i).Value];
-                }
+                    long index = rowIndex;
+                    if (invertMapIndices)
+                        index = longMapIndices.Length - 1 - index;
+                    if (index < setBufferMinRange || index >= setBufferMaxRange)
+                    {
+                        int bufferIndex = (int)(index / maxCapacity);
+                        setBuffer = ret._stringBuffers[bufferIndex];
+                        setBufferMinRange = bufferIndex * maxCapacity;
+                        setBufferMaxRange = (bufferIndex + 1) * maxCapacity;
+                    }
+                    index -= setBufferMinRange;
+                    if (mapIndex == null)
```

Comment:
Should you check for this first? No reason to do the above if `mapIndex` is null.

Reply:
Unfortunately no, I did try. The next line depends on `index` which depends on `setBufferMinRange` which is set in the lines above.